### ISSUE TITLE
Notification Center

### ISSUE THREAD

**Issue by MoeNick** [Giveth/giveth-dapps-v2#475]

**Comment by mohammadranjbarz** [comment]
@CarlosQ96  @aminlatifi 
I write anything comes to my mind, We can create backend issues when we finalized these items

[![](https://mermaid.ink/img/pako:eNqNk8FuwjAMhl_Fyhn1AXoeSNMQmgZol168xgVvrdMljhBiffelFUxFUI1IVWP7i2v7b06mdJZMbsg_Me48NoXA3ypkbK2ccsUlKjuB0zjSLxalHXmIgfyzhcXLFNB690ml3mWCepYd6LElKMyZXAZWsvBzObmVmsOVZ8lfg5llWWGmcrLWKen9oKVQem6HziYQapDrtaLGkEo7IGvyLpxfk_SfDiSaXuJWRHbjeu9tKR_O1YQCHN4IbUrzTiCJB1ZQB2HvDtA_DcoRogyQjKYeYAikCcMew3X-7jHhXj1V5ElKCg9omErcCn9Hmu7lSqR5P6R_yIt8j7CDsDdgByORRgEzMw35RNv0Qw_NFUb31KTq87S1VGGstW-lS2hsLSrNLavzJq-wDjQzGNWtj1KaXH2kC3S-Gmeq-wUAUvab)](https://mermaid-js.github.io/mermaid-live-editor/edit#pako:eNqNk8FuwjAMhl_Fyhn1AXoeSNMQmgZol168xgVvrdMljhBiffelFUxFUI1IVWP7i2v7b06mdJZMbsg_Me48NoXA3ypkbK2ccsUlKjuB0zjSLxalHXmIgfyzhcXLFNB690ml3mWCepYd6LElKMyZXAZWsvBzObmVmsOVZ8lfg5llWWGmcrLWKen9oKVQem6HziYQapDrtaLGkEo7IGvyLpxfk_SfDiSaXuJWRHbjeu9tKR_O1YQCHN4IbUrzTiCJB1ZQB2HvDtA_DcoRogyQjKYeYAikCcMew3X-7jHhXj1V5ElKCg9omErcCn9Hmu7lSqR5P6R_yIt8j7CDsDdgByORRgEzMw35RNv0Qw_NFUb31KTq87S1VGGstW-lS2hsLSrNLavzJq-wDjQzGNWtj1KaXH2kC3S-Gmeq-wUAUvab)

* We can have a table for **accountPreferences**,  then we put preferences of users there or we can save in user's entity ( I suggest creating different table and restrict reading records to just owner user)
* We should have an update endpoint to can change this preferences
* For every event we create insert an entity in **Notification** table
* Implement an endpoint for user to get their notifications
* Implement an endpoint to mark notification as read by user 
* Based on user's preferences we decide to send email to users for this notification or not, and we save email status in notification entity as **emailStatus**

@MoeNick 
Do you think these stuff meet your needs?

**Comment by CarlosQ96** [comment]
Good start! I would change the NotificationPreference table tho, because if a new notification is added we would need to migrate it and add a new row.
I would do the following:
1. Make the Notification table as general as posible (no projectId or userId) Just the types of notifications, title, descriptions. 
2. Every time we create a notification in this table, it would run a job that add new records to the **NotificationsPreferences** table. This table will have the notificationId and userId only and a boolean called Active or an Status (Send, Don't send, delay).  (This table will be by far the largests of all as it will have many records per user per notification type).
3. Every time a notification is going to be sent, we check the preference, create a new record in a new table called **NotificationToBeSent** that has the userId, NotificationId and the status (sent, NoNeedToSend and waitingforSend), and whether is read or not (boolean).
4. So the endpoint on item 4, would fetch from the table **NotificationToBeSent**.

Basically I changed a bit items, 1,3,4. And we would have 3 tables instead of 2.

Writing a diagram an will post it in a bit.

**Comment by CarlosQ96** [comment]
[![](https://mermaid.ink/img/pako:eNq9k81qHDEMx19F-NJL2AeYY2gCpSGUbkouc1FszY4Sj7215S5Ddt-9spmWaZK51jCMJf0RP3341djoyHSG0mfGQ8KpD_D33EfhgS0KxwCv60g9WRKHA8h8JOjNMcVnsnLHWcjBGRb7R_BvPHf80szdbtebrZwsXpN-HHSUbeJjhdpQ6K9YKalyxSZE38FzjuEaTiwjyEgQnyoNYHDwQnOGRD8LJ0WTCJnUW0U0IXugYVAp_yI__4t8WRtbnfuWaKBEwVJ-30QOQgdKUDKlLw5uv24Jwirjh8KldkDbxtWbVsQZXAyfBBYj25Fc8fSmjAq_hf8Qr2lPQf4He2v3XlBK1gJOyKLu25j2C33FOGu6eyL3EKv3_Qo9xegJA3D-Tug0zSNBUD2wtNGO8QT1mzDMUEITrQEztIDWBCPmzXmv3ObKTJQU3elDal3qjS7PpF3u9OpowOKlZrqotBwdCt04lphMN6DPdGWwSNzPwZpOV5f-iJYnuaguvwHIch9A)](https://mermaid-js.github.io/mermaid-live-editor/edit#pako:eNq9k81qHDEMx19F-NJL2AeYY2gCpSGUbkouc1FszY4Sj7215S5Ddt-9spmWaZK51jCMJf0RP3341djoyHSG0mfGQ8KpD_D33EfhgS0KxwCv60g9WRKHA8h8JOjNMcVnsnLHWcjBGRb7R_BvPHf80szdbtebrZwsXpN-HHSUbeJjhdpQ6K9YKalyxSZE38FzjuEaTiwjyEgQnyoNYHDwQnOGRD8LJ0WTCJnUW0U0IXugYVAp_yI__4t8WRtbnfuWaKBEwVJ-30QOQgdKUDKlLw5uv24Jwirjh8KldkDbxtWbVsQZXAyfBBYj25Fc8fSmjAq_hf8Qr2lPQf4He2v3XlBK1gJOyKLu25j2C33FOGu6eyL3EKv3_Qo9xegJA3D-Tug0zSNBUD2wtNGO8QT1mzDMUEITrQEztIDWBCPmzXmv3ObKTJQU3elDal3qjS7PpF3u9OpowOKlZrqotBwdCt04lphMN6DPdGWwSNzPwZpOV5f-iJYnuaguvwHIch9A)

**Comment by CarlosQ96** [comment]
@mohammadranjbarz @aminlatifi  let me know what you think.

**Comment by CarlosQ96** [comment]
Updated the diagram I made with the projectId, in the dev call I noticed I missed, people can subscribe to a project to receive their notifications.

**Comment by aminlatifi** [comment]
> * ****

The notification template is a great idea. We can define templates and in code simply ask notification server create a notification with a template and variables which fill the placeholders.

**Comment by mohammadranjbarz** [comment]
> [![](https://camo.githubusercontent.com/27dcb788c99914f17ce1fa8d60e1dd91859e1cc9aabaa00f76b7424c6703f015/68747470733a2f2f6d65726d6169642e696e6b2f696d672f70616b6f3a654e724e6b384675327a414d686c2d463047575849675f67593745564746595551394e68463139556959375a796d496d5551754d4a75382d53764147723631336e6744444a7657425f6e2d4b656a474f505a724f595070493970447331456634732d35596143426e68546a437933716e7269794a3467466b50694c3035706a34435a3363556862306349596c5f6862447138777450626477743976315a71736d53644369373239367a43375273597261495052566e4a52556458454462656a674b584f386868504a4344496938474e56417a5a36654d593551384966685a4a4b4534614d6d713051547059433444416f536a38787a48394c7671794472633539545468677775677776323069526345444a69675a303263504e312d32674b56375f32546936715f76676b745f774c7032704c317052735f674f58345157494c735276516c34437572316543577851652d786a31472d565f3874575062693557533165544a6b6d6a3668744e2d6356696c6e72586348614a5f344a70394f3471507a414674424d7233614c32572d5934516c5165534e69496a6e36412d6b34307a6c4e69677463414d62554e3977326a7a357479733075624b544a68557574634c3254725a4778334353552d6930302d50677931426171574c6f75586f7265416e54384c4a64494d4e47612d4d4c634c374f54725436525841333942797452667138677338456a4d61)](https://mermaid-js.github.io/mermaid-live-editor/edit#pako:eNrNk8Fu2zAMhl-F0GWXIg_gY7EVGFYUQ9NhF19UiY7ZymImUQuMJu8-SvAGr613ngDDJvWB_n-KejGOPZrOYPpI9pDs1Ef4s-5YaCBnhTjCy3qnriyJ4gFkPiL05pj4CZ3cUhb0cIYl_hbDq8wtPbdwt9v1ZqsmSdCi7296zC7RsYraIPRVnJRUdXEDbejgKXO8hhPJCDIi8GNVAzZ6eMY5Q8IfhZJKE4aMmq0QTpYC4DAoSj8xzH9LvqyDrc59TThgwugwv20iRcEDJigZ02cPN1-2gKV7_2Ti6q_vgkt_wLp2pL1pRs_gOX4QWILsRvQl4Cur1eCWxQe-xj1G-V_8tWPbi5WS1eTJkmj6htN-cVilnrXcHaJ_4Jp9O4qPzAFtBMr3aL2W-Y4QlQeSNiIjn6A-k40zlNigtcAMbUN9w2jz5tys0ubKTJhUutcL2TrZGx3CSU-i00-Pgy1BaqWLouXoreAnT8LJdIMNGa-MLcL7OTrT6RXA39BytRfq8gs8EjMa)

Thanks @CarlosQ96  your suggestions was great , I had not thought abouth those

**Comment by mohammadranjbarz** [comment]
@MoeNick  @CarlosQ96  List of notifications tha moenick gave me
https://docs.google.com/document/d/1gqotb3D2FG3yu8xEClhPXKXnWkwxfwDSpw5Eh54sv78/edit?usp=sharing

**Comment by MoeNick** [comment]
Under construction 😛

**Comment by CarlosQ96** [comment]
I changed the db structure for more flexibility a bit. Persisted the userWallet in its own entity and separated notification template from the data itself.
